chapters/grid-search.qmd

```diff
@@ -384,7 +383,497 @@ A simple workaround for creating a space-filling design is to repeat the unique
+
+```{r}
+#| label: fig-nested
+#| echo: false
+#| fig-align: center
+#| out-width: 80%
+#| fig-width: 8
+#| fig-height: 4
+#| fig-cap: Each line corresponds to a complete resampled grid search that uses the inner resamples. The points indicate the estimates of the optimal learning rate as determined by the inner results. The panel on the right contrasts the Brier score for the inner assessment set with the outer assessment sets. 
+
+p_trend <- 
+  nest_grid_lines %>% 
+  ggplot(aes(learn_rate, mean, col = id)) + 
+  geom_line(show.legend = FALSE, linewidth = 1, alpha = 1 / 3) + 
+  scale_x_log10() +
+  labs(x = "Learning Rate", y = "Brier Score")  +
+  geom_point(data = nest_grid_best, show.legend = FALSE) + 
+  ylim(brier_range) +
+  theme(plot.margin = margin(r = 0, unit = "pt"))
+
+p_diff <- 
+  two_levels %>% 
+  ggplot(aes(level, brier)) + 
+  geom_boxplot() +
+  geom_point(alpha = 1 / 5) +
+  ylim(brier_range) +
+  labs(y = NULL, x = NULL) + 
+  theme(axis.text.y = element_blank(),  axis.ticks.y = element_blank(),
+        plot.margin = margin(l = 0, unit = "pt"))
+
+p_trend + p_diff +  plot_layout(widths = c(5, 1))
+```
+
+We can see that the resample-specific optimal learning rates vary but are within a consistent region. There are a few resamples that found the same optimal value. There is some variation in the y-axis too; different assessment sets produce different values. The standard error of these inner statistics, `r format(nest_grid_std_error, digits = 3, scientific = FALSE)`, is much smaller than the value of the conventional estimate (`r format(grid_brier$std_err, digits = 3, scientific = FALSE)`).
+
+The panel on the left shows boxplots of the Brier scores for the inner resamples and the corresponding outer resample estimates. The standard error of the outer resamples is very similar to the the level of noise in the conventional estimate.  
+
+In the end, the nested resampling estimate of the Brier score was estimated as `r signif(nest_grid_brier$mean, 3)`; a value very close to the single 10-fold cross-validation result shown in @fig-1d-boost.
+
+::: {.important-box}
+It is important to emphasize that nested resampling is for **verification, not optimization**. It provides a better estimate of our model optimization process; it does not replace it. 
+:::
+
+We might best communicate the results of nested resampling like this: 
+
+> We tuned the learning rate of our boosted classification tree using grid search and found that a rate of 10<sup>`r signif(log10(grid_brier$learn_rate), 3)`</sup> to be best. We think that this model has a corresponding Brier score, measured without optimization bias, of `r signif(nest_grid_brier$mean, 3)`. 
+
+We can look at the candidates produced by the inner resamples to understand the stability of the optimization process and potentially diagnose other issues. We _would not_ choose "the best" inner resampling result and move forward with its candidate value. 
```

Comment:
I think it would be good to explain why we shouldn't use the best inner resampling result.  But I can't quite come up with the explanation.

Reply:
I added a note on that.